rdkit/Chem/EnumerateStereoisomers.py

```diff
@@ -0,0 +1,246 @@
+
+    But we can change that behavior:
+    >>> opts = StereoEnumerationOptions(onlyUnassigned=False)
+    >>> isomers = tuple(EnumerateStereoisomers(m, options=opts))
+    >>> len(isomers)
+    16
+
+    Since the result is a generator, we can allow exploring at least parts of very
+    large result sets:
+    >>> m = Chem.MolFromSmiles('Br' + '[CH](Cl)' * 20 + 'F')
+    >>> opts = StereoEnumerationOptions(maxIsomers=0)
+    >>> isomers = EnumerateStereoisomers(m, options=opts)
+    >>> for x in range(5):
+    ...   print(Chem.MolToSmiles(next(isomers),isomericSmiles=True))
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)Br
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)Br
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)Br
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@H](Cl)Br
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)Br
+
+    Or randomly sample a small subset:
+    >>> m = Chem.MolFromSmiles('Br' + '[CH](Cl)' * 20 + 'F')
+    >>> opts = StereoEnumerationOptions(maxIsomers=3)
+    >>> isomers = EnumerateStereoisomers(m, options=opts)
+    >>> for smi in sorted(Chem.MolToSmiles(x, isomericSmiles=True) for x in isomers):
+    ...     print(smi)
+    ...
+    F[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@@H](Cl)Br
+    F[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)Br
+    F[C@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@@H](Cl)[C@H](Cl)[C@@H](Cl)Br
+    """
+
+    tm = Chem.Mol(m)
+    flippers = _getFlippers(tm, options)
+    nCenters = len(flippers)
+    if not nCenters:
+        yield tm
+        return
+
+    if (options.maxIsomers == 0 or
+        2**nCenters <= options.maxIsomers):
+        bitsource = _RangeBitsGenerator(nCenters)
+    else:
+        if options.rand is None:
+            # deterministic random seed invariant to input atom order
```

Comment:
the `hash` function in python3 does not, by design, return reproducible results. https://stackoverflow.com/questions/27522626/hash-function-in-python-3-3-returns-different-results-between-sessions

I think it may not be important though. Maybe it makes sense for the default behavior to be non-deterministic. That way if I call the function multiple times on the same large molecule I sample different parts of space.

Reply:
Interesting, didn't know this. Reading into it more deeply it's only "str, bytes and datetime objects" that have this caveat. Since we're using a tuple of tuples of integers, this appears to be deterministic in Python 3. Though I could imagine them rescinding this in later version of Python 3 as the reason for the non-determinism is 'security'. 

So I've added another test case to check for this determinism so if a future Python 3 version changes to make int and/or tuple non-deterministic, we should at least see it in our testing. (The doc test I told you was failing in Python 3 wasn't failing anymore after that fateful merge, pushed a doctest syntax tweak to hopefully make it fail again.)

I would prefer keeping the behavior as reproducible as possible to make debugging easier. Can easily imagine a downstream script using this before conformer generation and only 'sometimes' working based on the isomer chosen, that would certainly be frustrating for anyone to debug. 

Would prefer if we were even deterministic between Python 2 and 3, but getting that would require way more effort seemingly outside the scope of this function. So I think the option of allowing the user to specify their own entire subclass of random.Random gives the user sufficient ability to achieve that level of determinism if they deem it important.

Reply:
So a happy happenstance of porting from using Random.sample to Random.getrandbits, we get determinism between Python 2 and 3! 

> ~$ ./anaconda2/bin/python -c 'import random; print(random.Random(1).getrandbits(10))'
> 137
> ~$ ./anaconda/bin/python -c 'import random; print(random.Random(1).getrandbits(10))'
>137
> ~$ ./anaconda2/bin/python -c 'import random; print(random.Random(2).getrandbits(10))'
> 978
> ~$ ./anaconda/bin/python -c 'import random; print(random.Random(2).getrandbits(10))'
> 978

But randint is not:
> ~$ ./anaconda2/bin/python -c 'import random; print(random.Random(1).randint(0, 1024))'
> 137
> ~$ ./anaconda/bin/python -c 'import random; print(random.Random(1).randint(0, 1024))'
> 275

But random is close... very strange
> ~$ ./anaconda2/bin/python -c 'import random; print(random.Random(1).random())'
> 0.134364244112
> ~$ ./anaconda/bin/python -c 'import random; print(random.Random(1).random())'
> 0.13436424411240122
> ~$ ./anaconda2/bin/python -c 'import random; print(random.Random(2).random())'
> 0.956034271889
> ~$ ./anaconda/bin/python -c 'import random; print(random.Random(2).random())'
> 0.9560342718892494